File: server/src/routes/favorites.js
const express = require('express');
const { authenticate } = require('../middleware/auth');
const logger = require('../utils/logger');

const router = express.Router();

// In-memory storage for favorites (in production, use database)
const userFavorites = new Map();

// Get user's favorite channels
router.get('/channels',
  authenticate,
  async (req, res) => {
    try {
      const userId = req.user.id;
      const favorites = userFavorites.get(userId) || [];
      
      res.json({
        success: true,
        favorites
      });
    } catch (error) {
      logger.error('Error fetching favorites:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error'
      });
    }
  });

// Add channel to favorites
router.post('/channels/:channelId',
  authenticate,
  async (req, res) => {
    try {
      const { channelId } = req.params;
      const userId = req.user.id;
      const { channelData } = req.body;
      
      let favorites = userFavorites.get(userId) || [];
      
      // Check if already favorited
      if (favorites.some(fav => fav.id === channelId)) {
        return res.status(400).json({
          success: false,
          message: 'Channel already in favorites'
        });
      }
      
      // Add to favorites
      const favoriteChannel = {
        id: channelId,
        name: channelData?.name || 'Unknown Channel',
        subscriberCount: channelData?.subscriberCount || 0,
        thumbnailUrl: channelData?.thumbnailUrl,
        channelUrl: channelData?.channelUrl || `https://youtube.com/channel/${channelId}`,
        addedAt: new Date().toISOString(),
        lastAnalyzed: channelData?.lastAnalyzed,
        avgOutlierScore: channelData?.avgOutlierScore || 0,
        isFavorited: true
      };
      
      favorites.push(favoriteChannel);
      userFavorites.set(userId, favorites);
      
      res.json({
        success: true,
        message: 'Channel added to favorites',
        favorite: favoriteChannel
      });
    } catch (error) {
      logger.error('Error adding favorite:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error'
      });
    }
  });

// Remove channel from favorites
router.delete('/channels/:channelId',
  authenticate,
  async (req, res) => {
    try {
      const { channelId } = req.params;
      const userId = req.user.id;
      
      let favorites = userFavorites.get(userId) || [];
      const initialLength = favorites.length;
      
      favorites = favorites.filter(fav => fav.id !== channelId);
      
      if (favorites.length === initialLength) {
        return res.status(404).json({
          success: false,
          message: 'Channel not found in favorites'
        });
      }
      
      userFavorites.set(userId, favorites);
      
      res.json({
        success: true,
        message: 'Channel removed from favorites'
      });
    } catch (error) {
      logger.error('Error removing favorite:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error'
      });
    }
  });

// Check if channel is favorited
router.get('/channels/:channelId/status',
  authenticate,
  async (req, res) => {
    try {
      const { channelId } = req.params;
      const userId = req.user.id;
      
      const favorites = userFavorites.get(userId) || [];
      const isFavorited = favorites.some(fav => fav.id === channelId);
      
      res.json({
        success: true,
        isFavorited
      });
    } catch (error) {
      logger.error('Error checking favorite status:', error);
      res.status(500).json({
        success: false,
        message: 'Internal server error'
      });
    }
  });

module.exports = router;

File: server-python/src/routes/favorites.py
from flask import Blueprint, request, jsonify, g
from middleware.auth import authenticate
from middleware.rbac import require_permission
from utils.logger import logger

bp = Blueprint('favorites', __name__, url_prefix='/api/favorites')

# Placeholder for favorites service
# In a real implementation, this would interact with a database

# Mock data store for favorites
favorites = {}

@bp.route('/', methods=['GET'])
@authenticate
@require_permission('favorites:read')
def get_favorites():
    """
    Get user's favorite channels
    """
    try:
        user_id = g.user['id']
        user_favorites = favorites.get(user_id, [])
        return jsonify({'favorites': user_favorites}), 200
    except Exception as e:
        logger.error(f"Error fetching favorites: {e}")
        return jsonify({'error': 'Failed to fetch favorites'}), 500

@bp.route('/', methods=['POST'])
@authenticate
@require_permission('favorites:write')
def add_favorite():
    """
    Add a channel to favorites
    """
    data = request.get_json()
    if not data or 'channelId' not in data:
        return jsonify({'error': 'channelId is required'}), 400
    
    try:
        user_id = g.user['id']
        channel_id = data['channelId']
        
        if user_id not in favorites:
            favorites[user_id] = []
        
        # Check if already favorited
        if channel_id not in favorites[user_id]:
            favorites[user_id].append(channel_id)
        
        return jsonify({'message': 'Channel added to favorites'}), 200
    except Exception as e:
        logger.error(f"Error adding favorite: {e}")
        return jsonify({'error': 'Failed to add favorite'}), 500

@bp.route('/<string:channel_id>', methods=['DELETE'])
@authenticate
@require_permission('favorites:write')
def remove_favorite(channel_id):
    """
    Remove a channel from favorites
    """
    try:
        user_id = g.user['id']
        
        if user_id in favorites and channel_id in favorites[user_id]:
            favorites[user_id].remove(channel_id)
            return jsonify({'message': 'Channel removed from favorites'}), 200
        else:
            return jsonify({'error': 'Channel not in favorites'}), 404
    except Exception as e:
        logger.error(f"Error removing favorite: {e}")
        return jsonify({'error': 'Failed to remove favorite'}), 500

@bp.route('/<string:channel_id>', methods=['GET'])
@authenticate
@require_permission('favorites:read')
def check_favorite(channel_id):
    """
    Check if a channel is in favorites
    """
    try:
        user_id = g.user['id']
        is_favorite = (user_id in favorites and channel_id in favorites[user_id])
        return jsonify({'isFavorite': is_favorite}), 200
    except Exception as e:
        logger.error(f"Error checking favorite: {e}")
        return jsonify({'error': 'Failed to check favorite'}), 500